### PR TITLE
Added mailgun example.

### PR DESCRIPTION
**Pull request by jmdobry**
@jonparrott

### CODE REVIEW

File: appengine/mailgun/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+```yaml
+runtime: nodejs
+vm: true
+api_version: 1
+env_variables:
+  PORT: 8080
+```
+
+Create a `config.js` in the root of your application with the following
```
Comment (jonparrott) [inline appengine/mailgun/README.md]: All of the other GAE node examples will use environment variables. Can you use that for consistency or do you feel strongly about config.js?

File: appengine/mailgun/README.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+### Deploy
+
+For convenience, you can use an npm script to run the `gcloud` command. Add
+these lines to your `package.json` file:
+
+```json
+"scripts": {
+  "start": "node ./bin/www",
+  "deploy": "gcloud preview app deploy app.yaml --promote --project <your-project-id>"
```
Comment (jonparrott) [inline appengine/mailgun/README.md]: they don't have to specify the project id for the gcloud command, gcloud init sets it for you.

File: appengine/mailgun/app.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  var err = new Error('Not Found');
+  err.status = 404;
+  next(err);
+});
+
+// error handlers
+
+// development error handler
+// will print stacktrace
+if (app.get('env') === 'development') {
```
Comment (jonparrott) [inline appengine/mailgun/app.js]: doesn't express do this by default? None of the other samples have all of this error handling stuff and it seems to distract from the point.

File: appengine/mailgun/bin/www [hunk 4]
```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env node
```
Comment (jonparrott) [inline appengine/mailgun/bin/www]: I know this is part of the generator, but again, none of the other samples we'll be releasing use this. How strongly are you attached to this?

**Comment by jonparrott** [comment]
In general, this samples seems to have too much code that isn't related to what it's demonstrating. The part about sending mail is fine though.

File: appengine/mailgun/README.md (hunk 1 above)
Comment (jmdobry) [inline appengine/mailgun/README.md]: How would the environment variable be set? I think it's unsafe (and not a practice I want to encourage) to put keys/secrets into a file that gets committed to source control. `config.js` is ignored by git. The [Geddy sample](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/tree/master/appengine/geddy#prepare-the-app) also uses a git-ignored file:

Comment (jonparrott) [inline appengine/mailgun/README.md]: I'm really torn on this. We need to be consistent with the other samples, which all use environment variables.

File: appengine/mailgun/README.md (hunk 2 above)
Comment (jmdobry) [inline appengine/mailgun/README.md]: That's true, but this command can't fail (for lack of specifying a project). It doesn't have any dependency on them having set the gcloud project correctly, gives the user better insight into what's happening, and is consistent with the other `appengine/` samples.

File: appengine/mailgun/app.js (hunk 3 above)
Comment (jmdobry) [inline appengine/mailgun/app.js]: I don't know, I just copied the `appengine/express/` sample.

File: appengine/mailgun/bin/www (hunk 4 above)
Comment (jmdobry) [inline appengine/mailgun/bin/www]: I'm not attached, I just copied the `appengine/express/` sample.

File: appengine/mailgun/README.md (hunk 1 above)
Comment (jmdobry) [inline appengine/mailgun/README.md]: One of my major friction points on my initial project was: How do I provide config to my app in various environments? I couldn't set environment variables without committing config to my app's source control (not gonna do that), so I used a git-ignored config file.

If we want the little samples to be simpler by just committing config to source control, then for the samples that have config I think there should be a disclaimer that says something like "In a production app, here's how you might get config into your app: <link to article about it or link to built-for-production sample app, e.g. LabelCat>"

**Comment by jonparrott** [comment]
After our offline discussion, do you want to update this PR or start a new one with the more bare-bones sample?

**Comment by jmdobry** [comment]
I'm updating this PR right now

**Comment by jonparrott** [comment]
:+1:

**Comment by jmdobry** [comment]
Updated

File: appengine/mailgun/README.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+> [Mailgun](https://www.mailgun.com/): The Email Service For Developers
+
+### Sign up for Mailgun
+
+1. Sign up for a [Mailgun account](https://mailgun.com/signup).
+1. Add a [new domain](https://mailgun.com/app/domains).
+1. Find your API key in your new domain's settings.
+
+### Create a new Node.js app
```
Comment (jonparrott) [inline appengine/mailgun/README.md]: This section title doesn't match the content. Maybe remove this section and move its content up to the first section "Mailgun on Google App Engine".

File: appengine/mailgun/README.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
+## Mailgun on Google App Engine
```
Comment (jonparrott) [inline appengine/mailgun/README.md]: Let's be more specific: "Mailgun on Google App Engine with Node.js".

File: appengine/mailgun/README.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+### Configure
+
+Create an `app.yaml` in the root of your application with the following
+contents:
+
+```yaml
+runtime: nodejs
+vm: true
+api_version: 1
+env_variables:
```
Comment (jonparrott) [inline appengine/mailgun/README.md]: Remove everything above env_variables, they're subject to change.

File: appengine/mailgun/README.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+1. Find your API key in your new domain's settings.
+
+### Create a new Node.js app
+
+This example uses [Express.js](http://expressjs.com) and
+[node-mailgun](http://github.com/shz/node-mailgun).
+
+### Configure
+
+Create an `app.yaml` in the root of your application with the following
```
Comment (jonparrott) [inline appengine/mailgun/README.md]: `Create` should be `modify`.

File: appengine/mailgun/README.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+api_version: 1
+env_variables:
+  PORT: 8080
+  MAILGUN_API_KEY: <your-mailgun-api-key>
+```
+
+### Start the app locally
+
+```
+$ MAILGUN_API_KEY=<your-mailgun-api-key> npm start
```
Comment (jonparrott) [inline appengine/mailgun/README.md]: Be consistent with other samples:

```
$ export MAILGUN_API_KEY=<your-mailgun-api-key>
$ npm start
```

File: appengine/mailgun/README.md [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+```
+
+Now visit http://localhost:8080 and try sending yourself an email.
+
+When the app is deployed to Google Cloud Platform the `MAILGUN_API_KEY`
+environment variable will automatically be supplied to the app.
```
Comment (jonparrott) [inline appengine/mailgun/README.md]: Replace `will automatically be supplied to the app` with `will be set to the value specified in app.yaml`.

File: appengine/mailgun/app.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+var http = require('http');
+var express = require('express');
+var path = require('path');
+var bodyParser = require('body-parser');
+
+// [START setup]
+var Mailgun = require('mailgun').Mailgun;
+var mg = new Mailgun(process.env.MAILGUN_API_KEY);
+
+var app = express();
```
Comment (jonparrott) [inline appengine/mailgun/app.js]: You can put express's setup outside of the fence. I think it's safe to assume most node developers know how to initialize express.

File: appengine/mailgun/app.js [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+// [END setup]
+
+var link = 'https://github.com/GoogleCloudPlatform/nodejs-docs-samples/tree/master/appengine/mailgun'; // jshint ignore:line
```
Comment (jonparrott) [inline appengine/mailgun/app.js]: Eh.. remove this. It doesn't add a ton of value. People will already be discovering this sample from our docs, and I'd rather them be linked to our docs than directly to the sample.

File: appengine/mailgun/app.js [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+app.set('view engine', 'jade');
+
+// Parse form data
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({ extended: false }));
+
+// [START index]
+app.get('/', function(req, res) {
+  res.render('index', {
+    title: 'Hello World! Express.js + Mailgun on Google App Engine.'
```
Comment (jonparrott) [inline appengine/mailgun/app.js]: You reference this same string multiple times. Why not just put it in the template?

File: appengine/mailgun/app.js [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+          sent: true
+        });
+      }
+    }
+  );
+});
+// [END hello]
+
+// [START server]
+var server = http.createServer(app).listen(
```
Comment (jonparrott) [inline appengine/mailgun/app.js]: why not just `app.listen`?

File: appengine/mailgun/package.json [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+  "scripts": {
+    "start": "node app.js",
+    "deploy": "gcloud preview app deploy app.yaml --promote"
+  },
+  "dependencies": {
+    "body-parser": "^1.14.1",
+    "express": "^4.13.3",
+    "jade": "^1.11.0",
+    "mailgun": "^0.5.0",
+    "serve-favicon": "^2.3.0"
```
Comment (jonparrott) [inline appengine/mailgun/package.json]: Is this used?

File: appengine/mailgun/views/error.jade [hunk 16]
```diff
@@ -0,0 +1,19 @@
+// Copyright 2015, Google, Inc.
```
Comment (jonparrott) [inline appengine/mailgun/views/error.jade]: Is this necessary/used?

File: appengine/mailgun/views/layout.jade [hunk 17]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2015, Google, Inc.
```
Comment (jonparrott) [inline appengine/mailgun/views/layout.jade]: If there's only one template, just combine the two.

**Comment by jmdobry** [comment]
Changes made

File: appengine/mailgun/README.md [hunk 18]
```diff
@@ -0,0 +1,59 @@
+## Mailgun on Google App Engine with Node.js
+
+> [Mailgun](https://www.mailgun.com/): The Email Service For Developers
+
+This example uses [Express.js](http://expressjs.com) and
+[node-mailgun](http://github.com/shz/node-mailgun).
```
Comment (jonparrott) [inline appengine/mailgun/README.md]: One more nit: 

```
This sample application demonstrates how to use [Express.js](http://expressjs.com) and [node-mailgun](http://github.com/shz/node-mailgun) to send transactional email on [Google App Engine](https://cloud.google.com/appengine).
```

File: appengine/mailgun/README.md [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+
+### Sign up for Mailgun
+
+1. Sign up for a [Mailgun account](https://mailgun.com/signup).
+1. Add a [new domain](https://mailgun.com/app/domains).
+1. Find your API key in your new domain's settings.
+
+### Configure
+
+Add these lines to the `app.yaml` file in the root of your application:
```
Comment (jonparrott) [inline appengine/mailgun/README.md]: I just realized there isn't a sample app.yaml in this PR, please add a sample app.yaml. They should only have to put in their api key, they shouldn't have to create the whole thing from scratch.

File: appengine/mailgun/app.js [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+    req.body.email,
+    // Subject
+    'Hello World!',
+    // Body
+    'Mailgun on Google App Engine with Node.js',
+    servername,
+    options,
+    function (err) {
+      if (err) {
+        return next(new Error('Failed to send email!'));
```
Comment (jonparrott) [inline appengine/mailgun/app.js]: Why not just `return next(err)`?

File: appengine/mailgun/README.md (hunk 19 above)
Comment (jmdobry) [inline appengine/mailgun/README.md]: https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/mailgun/appengine/mailgun/app.yaml

File: appengine/mailgun/views/index.jade [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+doctype html
+html
+  head
+    title= title
+    link(rel='stylesheet', href='/stylesheets/style.css')
```
Comment (jonparrott) [inline appengine/mailgun/views/index.jade]: This is no stylesheet.

There is no spoon

**Pull request merged**